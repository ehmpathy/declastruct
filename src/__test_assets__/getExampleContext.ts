import { DomainEntity } from 'domain-objects';
import { VisualogicContext } from 'visualogic';

import { getRef } from '..';
import {
  DeclaredResourceReference,
  DeclaredResourceReferenceKeyType,
} from '../domain/DeclaredResourceReference';
import { DeclastructContext } from '../domain/DeclastructContext';
import { DeclastructProvider } from '../domain/DeclastructProvider';
import { DeclastructProviderResourceRemoteStateInterface } from '../domain/DeclastructProviderResourceRemoteStateInterface';

const log = console;

interface CNCMachine {
  uuid?: string;
  serialNumber: string;
  size: string;
  location: string;
}
class CNCMachine extends DomainEntity<CNCMachine> implements CNCMachine {
  public static primary = ['uuid'] as const;
  public static unique = ['serialNumber'];
}
export type CNCMachineReference = DeclaredResourceReference<
  CNCMachine,
  'uuid',
  'serialNumber'
>;

interface CNCMachineRegistration {
  uuid?: string;
  machine: CNCMachineReference;
  registeredBy: string;
}
class CNCMachineRegistration
  extends DomainEntity<CNCMachineRegistration>
  implements CNCMachineRegistration
{
  public static primary = ['uuid'] as const;
  public static unique = ['machine'];
}
export type CNCMachineRegistrationReference = DeclaredResourceReference<
  CNCMachineRegistration,
  'uuid',
  'machine'
>;

const exampleMachine = new CNCMachine({
  uuid: 'de0e34ca-ce61-4884-ace5-093a1a32ff92',
  serialNumber: '821',
  size: 'big',
  location: 'basement',
}) as Required<CNCMachine>;

const exampleRegistration = new CNCMachineRegistration({
  uuid: '9857cc32-8bcf-4291-8850-aef94c2081f9',
  machine: getRef(exampleMachine),
  registeredBy: 'bob',
}) as Required<CNCMachineRegistration>;

const cncMachineRemoteStateInterface: DeclastructProviderResourceRemoteStateInterface<
  CNCMachine,
  VisualogicContext,
  'uuid',
  'serialNumber'
> = {
  for: CNCMachine.name,
  findByPrimary: async ({ uuid }) => {
    if (uuid === exampleMachine.uuid) return exampleMachine;
    return null;
  },
  findByUnique: async ({ serialNumber }) => {
    if (serialNumber === exampleMachine.serialNumber) return exampleMachine;
    return null;
  },
  create: () => Promise.reject(new Error('todo')),
  update: () => Promise.reject(new Error('todo')),
  destroy: () => Promise.reject(new Error('todo')),
};

const cncMachineRegistrationRemoteStateInterface: DeclastructProviderResourceRemoteStateInterface<
  CNCMachineRegistration,
  VisualogicContext,
  'uuid',
  'machine'
> = {
  for: CNCMachineRegistration.name,
  findByPrimary: async ({ uuid }) => {
    if (uuid === exampleRegistration.uuid) return exampleRegistration;
    return null;
  },
  findByUnique: async ({ machine: machineRef }) => {
    if (
      machineRef.identifiedBy.key ===
        DeclaredResourceReferenceKeyType.PRIMARY_KEY &&
      machineRef.identifiedBy.value.uuid === exampleMachine.uuid
    )
      return exampleRegistration;
    if (
      machineRef.identifiedBy.key ===
        DeclaredResourceReferenceKeyType.UNIQUE_KEY &&
      machineRef.identifiedBy.value.serialNumber === exampleMachine.serialNumber
    )
      return exampleRegistration;
    return null;
  },
  create: () => Promise.reject(new Error('todo')),
  destroy: () => Promise.reject(new Error('todo')),
};

const exampleContext: DeclastructContext & VisualogicContext = {
  log,
  providers: [
    {
      agentOptions: {},
      interfaces: {
        [cncMachineRemoteStateInterface.for]: cncMachineRemoteStateInterface,
        [cncMachineRegistrationRemoteStateInterface.for]:
          cncMachineRegistrationRemoteStateInterface,
      } as any as DeclastructProvider<any>['interfaces'], // TODO: fix types to make this assertion not needed
      hooks: {},
    },
  ],
  resources: [], // no resources needed in context for this test
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getExampleContext = () => ({
  exampleContext,
  exampleMachine,
  CNCMachine,
  exampleRegistration,
  CNCMachineRegistration,
});
