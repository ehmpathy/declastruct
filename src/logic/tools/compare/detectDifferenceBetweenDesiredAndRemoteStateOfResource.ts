import { DetailedDiff, detailedDiff as objectDiff } from 'deep-object-diff';
import { DomainObject, omitMetadataValues } from 'domain-objects';
import { diff as jestDiff } from 'jest-diff';
import { VisualogicContext } from 'visualogic';

import { DeclaredResource } from '../../../domain/DeclaredResource';
import { DeclaredResourceReference } from '../../../domain/DeclaredResourceReference';
import { DeclastructContext } from '../../../domain/DeclastructContext';
import { castReferenceToGrokableString } from './castReferenceToGrokableString';

/**
 * recursively traverse each key of the object and grokify any references found
 */
const withReferencesGrokified = async (
  { resource }: { resource: DeclaredResource },
  context: DeclastructContext & VisualogicContext,
) => {
  // create a clone of the resources, which we'll mutate keys on
  const withKeysGrokified: Record<string, any> =
    new (resource.constructor as typeof DomainObject)(resource);

  // for each key holding a reference, replace reference with grokable string
  await Promise.all(
    Object.keys(resource).map(async (key) => {
      const value = (resource as any)[key];
      if (value instanceof DeclaredResourceReference)
        withKeysGrokified[key] = await castReferenceToGrokableString(
          { reference: value },
          context,
        );
      else if (Array.isArray(value))
        withKeysGrokified[key] = await Promise.all(
          value.map(async (valueItem) =>
            value instanceof DeclaredResourceReference
              ? await castReferenceToGrokableString(
                  { reference: valueItem },
                  context,
                )
              : await withReferencesGrokified({ resource: valueItem }, context),
          ),
        );
      else if (value instanceof DomainObject)
        withKeysGrokified[key] = await withReferencesGrokified(
          { resource: value },
          context,
        );
    }),
  );

  // return the fingerprinted object
  return withKeysGrokified;
};

export const detectDifferenceBetweenDesiredAndRemoteStateOfResource = async <
  T extends DomainObject<any>,
>(
  {
    desiredState,
    remoteState,
  }: {
    desiredState: T | null;
    remoteState: T | null;
  },
  context: DeclastructContext & VisualogicContext,
): Promise<{
  displayable: string;
  usable: DetailedDiff;
}> => {
  // replace all nested references with reference fingerprints, for readability
  const diffableDesiredState = desiredState
    ? // remove the metadata values from the desired state for diff, since these properties dont describe the data, they are just additional data typically generated by the remote state which cant be locally declared (e.g., uuid, createdAt, etc)
      omitMetadataValues(
        await withReferencesGrokified({ resource: desiredState }, context),
      )
    : null;
  const diffableRemoteState = remoteState
    ? // remove the metadata values from the remote state for diff, since these properties dont describe the data, they are just additional data typically generated by the remote state which cant be locally declared (e.g., uuid, createdAt, etc)
      omitMetadataValues(
        await withReferencesGrokified({ resource: remoteState }, context),
      )
    : null;

  // define a displayable difference
  const displayableDifference: string = jestDiff(
    diffableDesiredState ?? {},
    diffableRemoteState ?? {},
    {
      omitAnnotationLines: true,
      aAnnotation: 'Desired State',
      bAnnotation: 'Remote State',
      aIndicator: '+',
      bIndicator: '-',
    },
  )!;

  // replace the "Object {}" placeholder we have to use to represent null state, if present (since jest shows "Comparing two different types of values. Expected object but received null." if we try to use null directly)
  const cleanedDisplayableDifference = displayableDifference
    ?.replace('- Object {}', '- null')
    .replace('+ Object {}', '+ null');

  // define a usable difference
  const usableDifference = objectDiff(
    diffableRemoteState ?? {},
    diffableDesiredState ?? {},
  ); // TODO: improve the usable difference. use the `objectDiff` and add our own info + make a more useful type

  // return both
  return {
    displayable: cleanedDisplayableDifference,
    usable: usableDifference,
  };
};
