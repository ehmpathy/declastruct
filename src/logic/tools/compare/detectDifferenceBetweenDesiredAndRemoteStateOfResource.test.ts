import { getExampleContext } from '../../../__test_assets__/getExampleContext';
import { detectDifferenceBetweenDesiredAndRemoteStateOfResource } from './detectDifferenceBetweenDesiredAndRemoteStateOfResource';

const { exampleContext, exampleMachine, CNCMachine } = getExampleContext();

describe('detectDifferenceBetweenDesiredAndRemoteStateOfResource', () => {
  it('should return a helpful displayable difference and usable difference', async () => {
    const desiredState = exampleMachine;
    const remoteState = new CNCMachine({
      ...desiredState,
      location: 'moon',
    });
    const difference =
      await detectDifferenceBetweenDesiredAndRemoteStateOfResource(
        {
          desiredState,
          remoteState,
        },
        exampleContext,
      );
    expect(difference).toMatchSnapshot();
  });
  it('should show the case where we want to create a new resource well', async () => {
    const desiredState = exampleMachine;
    const remoteState = null;
    const difference =
      await detectDifferenceBetweenDesiredAndRemoteStateOfResource(
        {
          desiredState,
          remoteState,
        },
        exampleContext,
      );
    console.log(difference.displayable);
    console.log(difference.usable);
    expect(difference).toMatchSnapshot();
  });
});
