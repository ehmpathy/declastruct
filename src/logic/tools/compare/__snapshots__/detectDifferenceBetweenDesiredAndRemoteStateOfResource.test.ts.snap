// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detectDifferenceBetweenDesiredAndRemoteStateOfResource should return a helpful displayable difference and usable difference 1`] = `
{
  "displayable": "[2m  CNCMachine {[22m
[32m+   "location": "basement",[39m
[31m-   "location": "moon",[39m
[2m    "serialNumber": "821",[22m
[2m    "size": "big",[22m
[2m  }[22m",
  "usable": {
    "added": {},
    "deleted": {},
    "updated": {
      "location": "basement",
    },
  },
}
`;

exports[`detectDifferenceBetweenDesiredAndRemoteStateOfResource should show the case where we want to create a new resource well 1`] = `
{
  "displayable": "[32m+ CNCMachine {[39m
[32m+   "location": "basement",[39m
[32m+   "serialNumber": "821",[39m
[32m+   "size": "big",[39m
[32m+ }[39m
[31m- null[39m",
  "usable": {
    "added": {
      "location": "basement",
      "serialNumber": "821",
      "size": "big",
    },
    "deleted": {},
    "updated": {},
  },
}
`;
